### ISSUE TITLE
TaskLocal and Local quarantine function

### ISSUE THREAD

**Issue by mihaisoloi** [monix/monix#837]
When calling third-party APIs you have no guarantee that it's not doing things that make you lose your Local instance, so I propose a function that does something similar to this:

```scala
  val globalLocal = Local[Int](0)
  def quarantine[A](task: => Task[A], scheduler: Scheduler): Task[A] = {
    val old = globalLocal.get
    task.asyncBoundary(scheduler).map { a =>
      globalLocal.update(old)
      a
    }
  }
```
This way we get to keep the local instance. 

I've written a `TaskLocalApp` that makes use of the above function using a `CompletableFuture` from java-8 which looses the `Local` instance.
https://gist.github.com/mihaisoloi/04f9ddbb16292fa3b6b3769d2744244e

**Comment by oleg-py** [comment]
If you don't specify Executor for Java methods, it would use `ForkJoinPool.commonPool()` which does not support tracing.
I'm pretty sure that `toScala` converter you're using _also_ does that. Maybe consider trying it with `Task.fromFutureLike`?

**Comment by mihaisoloi** [comment]
@oleg-py I've tried it with `Task.fromFutureLike` and it still doesn't propagate the context, even if I specify an executor that supports tracing the context is lost.

**Comment by Avasil** [comment]
Thanks for detailed report. :)

I'll try to investigate in the next couple of days

**Comment by mdedetrich** [comment]
Isn't `monix.execution.schedulers.TracingScheduler` meant to solve this problem, or am I missing something?

**Comment by Avasil** [comment]
@mihaisoloi 
Is it possible that the issue is because you're using `val` in the `Future` so it starts evaluating earlier?
Following works for me:
```scala
  def futureLike = CompletableFuture.runAsync(() => {
    println(s"Starting futureLike ${Thread.currentThread().getName}")
    TimeUnit.SECONDS.sleep(1)
    println(
      s"Future like on thread: ${Thread.currentThread.getName} local: ${globalLocal.get}")
  }, io)
```


If #839 gets merged it seems like it also fixes this behaviour in case it's `val`:
```scala
Starts on thread: scala-execution-context-global-12 local: 0
Future like on thread: ForkJoinPool.commonPool-worker-1 local: 0 // this should be 1 but it's not because it looses Local
Ends on thread: ForkJoinPool.commonPool-worker-1 local: 0 // this should be 1
```

it ends on `1` without `quarantine`.

**Comment by Avasil** [comment]
@mihaisoloi You can try with `3.0.0-RC2-SNAPSHOT-9e79718-SNAPSHOT`

**Comment by ljwagerfield** [comment]
@Avasil 
> Is it possible that the issue is because you're using val to create the Future so it starts evaluating earlier?

But what if the `CompletableFuture` was created by some third-party library which uses it's own threading model, that doesn't even use `Executor`?

For example, if you called `client.makeRequest(): CompletableFuture` on a Play WS Client, it would construct a `new CompletableFuture()` which would manually be marked as "completed" during an epoch on some Netty event loop.

In a situation like this, how can you ensure that the continuations that follow the `makeRequest` method have the correct `Local` in scope?

**Comment by mdedetrich** [comment]
This ticket might be related to https://github.com/monix/monix/issues/848

Note that in my case as I have set up the test (see https://github.com/mdedetrich/monix-mdc/blob/master/src/test/scala/MDCFutureSpec.scala) to use `TracingScheduler` (which according to documentation should pass the `TaskLocal` along each execution step when using the `TracingScheduler` as the `Future`'s `ExecutionContext`.

Also note that `3.0.0-RC2-SNAPSHOT-9e79718-SNAPSHOT` is being used

@mihaisoloi Are you using the `TracingScheduler`? This is meant to be the way to make sure the `TaskLocal` is shared properly with `Future`.

**Comment by ljwagerfield** [comment]
I've written the following test...

It highlights the problem of executing blackbox asynchronous code, which may be using a custom threading model that does not use an `Executor`:

```scala
import java.util.concurrent.CompletableFuture

import monix.eval.Task
import monix.execution.Scheduler
import monix.execution.misc.Local
import scala.compat.java8.FutureConverters.toScala
import cats.implicits._

object Application extends App {
  def blackbox(): CompletableFuture[Unit] = {
    val f = new CompletableFuture[Unit]()
    val t = new Thread {
      override def run(): Unit = {
        Thread.sleep(1000)
        f.complete(())
      }
    }
    t.start()
    f
  }

  implicit val scheduler: Scheduler = Scheduler.traced

  val local       = Local[Int](0)
  val increment   = Task.eval { local.update(local() + 1) }
  val assertion   = Task.eval { assert(local() == 2, s"Expected '2' but got '${local()}'.") }
  val runBlackbox = Task.fromFuture(toScala(blackbox()))
  val test        = increment *> runBlackbox *> increment *> assertion

  test.runSyncUnsafe()
}
```

This test currently fails with `3.0.0-RC2` -- AFAIK I haven't seen any discussions around fixing cases like this?

**Comment by mihaisoloi** [comment]
@mdedetrich @Avasil I think @ljwagerfield hit the nail on the head with the code example he pasted, this is what I was trying to achieve

**Comment by Avasil** [comment]
@ljwagerfield @mihaisoloi It might be related to: #848

**Comment by Avasil** [comment]
@ljwagerfield do you have use cases for code like this? Is it used internally in the libraries that you need to work with? And is `Local` modified there or is it enough to have something like `quarantine` there (preferably under the hood) ?

**Comment by ljwagerfield** [comment]
> Is it used internally in the libraries that you need to work with?

Yes. We have a third-party library that uses Netty internally: it returns a `CompletableFuture` to us, but internally it will complete this on an epoch in the Netty event loop.

>  And is Local modified there?

No. We modify `Local` before the call, and attempt to read it after the call (at which point it's been lost).

> is it enough to have something like quarantine there?

Yes. After we call the third-party method, we call `quarantine` on the resulting `Task`. This ensures `Local` is restored correctly after the call completes.

**Example:**
```scala
implicit class RichTask[A](val value: Task[A]) extends AnyVal {
  def quarantine: Task[A] =
    for {
      old    <- Task(Local.getContext()) // Must be wrapped in 'Task()' to prevent immediate/premature 'get'.
      result <- value.asyncBoundary // Shift back to default execution context (from whatever the untrusted 'task' left us on).
      _      <- Task(Local.setContext(old))
    } yield result
}

object App {
  val thirdPartyLibrary = ???
  val fooUnsafe: CompletableFuture[Int] = thirdPartyLibrary.fooAsync()
  val fooSafe: Task[Int] = Task.from(fooUnsafe).quarantine

  // ...now use 'fooSafe' without any problems...
}
```

**Comment by Avasil** [comment]
@mihaisoloi @ljwagerfield Can you check on `3.0.0-RC2-9c49e02-SNAPSHOT`? It includes changes from #866

**Comment by oleg-py** [comment]
Since we already merged #887, where we have `{Task}Local.isolate`, I'll go ahead and close this issue.
